pilot/pkg/serviceregistry/cloudfoundry/servicediscovery.go

```diff
@@ -101,6 +120,38 @@ func (sd *ServiceDiscovery) Instances(hostname model.Hostname, ports []string, t
 		})
 	}
 
+	internalRoutesResp, err := sd.Client.InternalRoutes(context.Background(), new(copilotapi.InternalRoutesRequest))
+	if err != nil {
+		return nil, fmt.Errorf("getting internal routes: %s", err)
+	}
+
+	internalRouteServicePort := &model.Port{
+		Port:     sd.ServicePort,
+		Protocol: model.ProtocolTCP,
+		Name:     "tcp",
+	}
+
+	for _, internalRoute := range internalRoutesResp.GetInternalRoutes() {
+		for _, backend := range internalRoute.GetBackends().Backends {
+			if internalRoute.Hostname == hostname.String() {
+				instances = append(instances, &model.ServiceInstance{
```

Comment:
Don’t you want to dedup with ones added in for loop above?

Reply:
Routes and InternalRoutes are mutually exclusive. This is handled by Cloud Foundry's Copilot, which filters on `apps.internal` domain right now.